### PR TITLE
fix(compiler): use either summary or metadata information when reading from `.d.ts`

### PR DESCRIPTION
**Pull request by tbosch**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: packages/compiler/src/aot/static_symbol_resolver.ts [hunk 1]
```diff
@@ -251,17 +251,20 @@ export class StaticSymbolResolver {
...
-    // have summaries, only .d.ts files. So we always need to check both, the summary
-    // and metadata.
-    let symbols = new Set<StaticSymbol>(this.summaryResolver.getSymbolsOf(filePath));
+    // have summaries, only .d.ts files, but `summaryResolver.isLibraryFile` returns true.
     this._createSymbolsOf(filePath);
+    const metadataSymbols: StaticSymbol[] = [];
     this.resolvedSymbols.forEach((resolvedSymbol) => {
       if (resolvedSymbol.symbol.filePath === filePath) {
-        symbols.add(resolvedSymbol.symbol);
+        metadataSymbols.push(resolvedSymbol.symbol);
```
Comment (chuckjaz) [inline packages/compiler/src/aot/static_symbol_resolver.ts]: The use of the set was to ensure this array has no duplicates. Is that now ensured some other way?

File: packages/compiler/src/aot/summary_resolver.ts [hunk 2]
```diff
@@ -85,11 +87,12 @@ export class AotSummaryResolver implements SummaryResolver<StaticSymbol> {
...
 
-  private _loadSummaryFile(filePath: string) {
-    if (this.loadedFilePaths.has(filePath)) {
-      return;
+  private _loadSummaryFile(filePath: string): boolean {
+    let hasSummary = this.loadedFilePaths.get(filePath);
+    if (hasSummary != null) {
+      return hasSummary;
     }
-    this.loadedFilePaths.add(filePath);
```
Comment (chuckjaz) [inline packages/compiler/src/aot/summary_resolver.ts]: With this moving to after the `if` could potentially allow indirect recursion that was prevented by seeding the cache. Is this is concern?

**Review by chuckjaz — COMMENTED** [review]
Some questions.

File: packages/compiler/src/aot/summary_resolver.ts (hunk 2 above)
Comment (tbosch) [reply packages/compiler/src/aot/summary_resolver.ts]: Good point. Will change that.

File: packages/compiler/src/aot/static_symbol_resolver.ts (hunk 1 above)
Comment (tbosch) [reply packages/compiler/src/aot/static_symbol_resolver.ts]: We needed the `Set` as we merged 2 list of symbols. As we are now always using either or, and we know that these lists don't contain duplicates, we are good.

File: packages/compiler/src/aot/summary_resolver.ts (hunk 2 above)
Comment (tbosch) [reply packages/compiler/src/aot/summary_resolver.ts]: (although this shouldn't happen...)

**Review by chuckjaz — APPROVED** [review]

**Comment by tbosch** [comment]
Travis is red because of a flake in AIO.

File: packages/bazel/src/ngc-wrapped/index.ts [hunk 3]
```diff
@@ -63,13 +63,30 @@ export function main(args) {
     ngOptions.annotationsAs = 'static fields';
   }
 
+  const rootDirs = tsOptions.rootDirs;
+  if (!rootDirs) {
+    throw new Error('rootDirs is not set!');
+  }
+  // When we canonicalize, we want to strip the longest rootDir.
+  rootDirs.sort((a, b) => b.length - a.length);
```
Comment (alexeagle) [inline packages/bazel/src/ngc-wrapped/index.ts]: I think it belongs in rules_typescript, let me move it there?

File: packages/bazel/src/ngc-wrapped/index.ts [hunk 4]
```diff
@@ -63,13 +63,30 @@ export function main(args) {
...
+
   function relativeToRootDir(filePath: string): string {
     if (tsOptions.rootDir) {
       const rel = path.relative(tsOptions.rootDir, filePath);
       if (rel.indexOf('.') != 0) return rel;
     }
     return filePath;
   }
+
+  function relativeToRootDirs(filePath: string): string {
```
Comment (alexeagle) [inline packages/bazel/src/ngc-wrapped/index.ts]: let's have one function that takes a parameter 
roots
that defaults to `[tsOptions.rootDir]`
to reduce the duplication?

**Review by alexeagle — APPROVED** [review]
last two commits

File: packages/bazel/src/ngc-wrapped/index.ts (hunk 3 above)
Comment (alexeagle) [reply packages/bazel/src/ngc-wrapped/index.ts]: actually this already happens in the compiler_host.ts from rules_typescript, shouldn't need it

**Pull request closed without merging**